### PR TITLE
Boxplot Closes #701

### LINKED ISSUE #701: Adding Box Plot series
This is a series popular among statisticians and different scientists:

https://en.wikipedia.org/wiki/Box_plot

I believe that we should support it (as we support candlesticks and other specialized series).

### PR DESCRIPTION
**Pull request by bjedrzejewski**
This is the boxplot pull request. Please review. There is no visual test written yet.

### CODE REVIEW

**Comment by bjedrzejewski** [comment]
image

File: site/src/components/series/boxPlot.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      .datum(data)
+      .call(boxPlot);
+
+---
+
+A [box plot series](https://en.wikipedia.org/wiki/Box_plot) is a convenient way of graphically depicting groups of 
+numerical data through their quartiles. Boxes can be renderer vertically or horizontally based on the value of the `orient` property.
+
+The upper and lower end of each box are defined by the `boxUpper` and `boxLower` properties.
+The upper and lower whisker of each box are defined by the `whiskerUpper` and `whiskerLower` properties.
```
Comment (ColinEberhardt) [inline site/src/components/series/boxPlot.md]: It's probably worth noting that when the bar is vertical the box / whisker properties are in the yScale coordinate system, whereas when it is horizontal, they are in the xScale coordinate system.

File: src/series/boxBase.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                y: yScale(yValue(d, i)),
+                boxHigh: xScale(boxHigh(d, i)),
+                boxLow: xScale(boxLow(d, i)),
+                whiskerHigh: xScale(whiskerHigh(d, i)),
+                whiskerLow: xScale(whiskerLow(d, i))
+            };
+        }
+    };
+    base.defined = function(d, i) {
+        return whiskerLow(d, i) != null && whiskerHigh(d, i) != null
```
Comment (ColinEberhardt) [inline src/series/boxBase.js]: How about adding a utility function that makes this a bit easier to manage?

```
allPropertiesDefined[whiskerHigh, whiskerLow, boxHigh, boxLow, xValue, yValue)
```

File: src/series/boxPlot.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import boxBase from './boxBase';
+
+export default function() {
+
+    var decorate = noop,
+        barWidth = 5,
+        orient = 'vertical',
+        base = boxBase();
+
+    var dataJoin = dataJoinUtil()
```
Comment (ColinEberhardt) [inline src/series/boxPlot.js]: @jdunkerley - just wondering if you made any more progress on your datajoin utilities? I'd love to see this repeated pattern eliminated!

File: src/series/boxPlot.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                    .y(values.y)
+                    .x(values.x)
+                    .boxHigh(values.boxHigh)
+                    .boxLow(values.boxLow)
+                    .whiskerHigh(values.whiskerHigh)
+                    .whiskerLow(values.whiskerLow);
+
+                gboxPlot.select('path')
+                    .attr('d', pathGenerator([d]))
+                    .attr('stroke', 'black');
```
Comment (ColinEberhardt) [inline src/series/boxPlot.js]: This should be a class

File: src/svg/boxPlot.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                whiskerToBotBox = boxLow(d, i) - whiskerLow(d, i),
+                boxToMid = y(d, i) - boxLow(d, i),
+                midToBox = boxHigh(d, i) - y(d, i),
+                boxToWhiskerHigh = whiskerHigh(d, i) - boxHigh(d, i),
+                yBottom = y(d, i) - whiskerLow(d, i),
+                yTop = whiskerHigh(d, i) - y(d, i),
+                xBottom = x(d, i) - whiskerLow(d, i),
+                xTop = whiskerHigh(d, i) - x(d, i);
+
+            var boxSerieVertical = '';
```
Comment (ColinEberhardt) [inline src/svg/boxPlot.js]: [nitpick] I think you're missing an 's' in the variable name.

File: src/svg/boxPlot.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+                var verticalBar = 'v' + (-halfWidth) + 'v' + (2 * halfWidth) + 'v' + (-halfWidth),
+                    horizontalToBotBox = 'h' + (-whiskerToBotBox),
+                    horizontalToMidBox = 'h' + (-boxToMid),
+                    horizontalToMidBoxDown = 'h' + (boxToMid),
+                    horizontalToTopBox = 'h' + (-midToBox),
+                    horizontalToTopBoxDown = 'h' + (midToBox),
+                    horizontalToWhisker = 'h' + (-boxToWhiskerHigh);
+
+                boxSerieHorizontal = 'M' + xBottom + ',0'
```
Comment (ColinEberhardt) [inline src/svg/boxPlot.js]: A path mini-language would help a lot here https://github.com/ScottLogic/d3fc/issues/480

**Comment by ColinEberhardt** [comment]
Just a few minor comments, but other than that, looks really good to me. 

One thing that probably requires a little thought is transitions, when a box-plot enters it should probably transition in the same way as OHLC / candlestick, i.e. grow from its centre-point. See #741

**Comment by bjedrzejewski** [comment]
Ok, will fix the code at some point this week! Thank you for review and comments.

**Comment by bjedrzejewski** [comment]
Hi @ColinEberhardt, sorry that it took so long, but I think I applied all your suggestions and this should be ready for merge/another review.

File: src/series/boxBase.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                y: yScale(yValue(d, i)),
+                boxHigh: xScale(boxHigh(d, i)),
+                boxLow: xScale(boxLow(d, i)),
+                whiskerHigh: xScale(whiskerHigh(d, i)),
+                whiskerLow: xScale(whiskerLow(d, i))
+            };
+        }
+    };
+    base.defined = function(d, i) {
+        return defined(d, i, whiskerLow, whiskerHigh, boxLow, boxHigh, xValue, yValue);
```
Comment (ColinEberhardt) [inline src/series/boxBase.js]: The purpose of the `defined` function is to determine whether the value accessors return suitable values for the given `d` and `i` arguments. I wouldn't expect `defined` to include the `d` and `i` arguments in its list.

See the similar implementation here:


https://github.com/ScottLogic/d3fc/blob/master/src/series/errorBase.js#L45

File: src/svg/boxPlot.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                whiskerToBotBox = boxLow(d, i) - whiskerLow(d, i),
+                boxToMid = y(d, i) - boxLow(d, i),
+                midToBox = boxHigh(d, i) - y(d, i),
+                boxToWhiskerHigh = whiskerHigh(d, i) - boxHigh(d, i),
+                yBottom = y(d, i) - whiskerLow(d, i),
+                yTop = whiskerHigh(d, i) - y(d, i),
+                xBottom = x(d, i) - whiskerLow(d, i),
+                xTop = whiskerHigh(d, i) - x(d, i);
+
+            var boxSeriesVertical = '';
```
Comment (ColinEberhardt) [inline src/svg/boxPlot.js]: considering the if / else block below - is there any reason why you have separate `boxSeriesVertical` and `boxSeriesHorizontal` variables? wouldn't a single `boxSeriesPath` variable be sufficient?

File: src/util/defined.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+// Checks that passes properties are 'defined', meaning that calling them with d, i returns non null values
+export default function(d, i) {
+    for (var c = 0, j = arguments.length; c < j; c++) {
+        if (arguments[c] !== d && arguments[c] !== i) {
```
Comment (ColinEberhardt) [inline src/util/defined.js]: Ah ... I can see why your code in `boxBase` passed `d` and `i`!

What this function should do is take a list of functions, returning a new function which you invoke like so:

```
function upper(d,i) { ... };
function lower(d,i) { ... };
function center(d,i) { ... };

defined(upper, lower, center)(d, i);
```

Does that make sense?

Considering the usage above, I'd move this to the `fc.util.fn` namespace.

**Comment by ColinEberhardt** [comment]
Thanks for the updates @bjedrzejewski - almost there! Just a few minor tweaks required.

One last thing, can you also roll-out usage of the new `defined` utility function in order to keep the series implementations consistent? Thanks.

File: src/util/defined.js (hunk 9 above)
Comment (bjedrzejewski) [inline src/util/defined.js]: It makes perfect sense, I will refactor, move it to fc.util.fn and roll-out the usage.

Comment (bjedrzejewski) [inline src/util/defined.js]: I am trying to do this:

`// Returns function that checks if passes properties are 'defined', meaning that calling them with (d, i) returns non null values
export function defined() {

    var self = this;

    return function(d, i) {
        for (var c = 0, j = self.arguments.length; c < j; c++) {
            if (self.arguments[c](d, i) === null) {
                return false;
            }
        }
    };
}`

But eslint stops me from `var self = this;` any suggestions?

Comment (bjedrzejewski) [inline src/util/defined.js]: Nevermind:

var givenArguments = this.arguments;

sorry :)

**Comment by bjedrzejewski** [comment]
@ColinEberhardt there is a major issue with using defined in that way:

`base.defined = defined(errorLow, errorHigh, xValue, yValue);`

Every time any of the `errorLow, errorHigh, xValue, yValue` are being redefined, the `base.defined = defined(errorLow, errorHigh, xValue, yValue);` should be called again. For example from the:

<code>
base.errorLow = function(x) {
        if (!arguments.length) {
            return errorLow;
        }
        errorLow = d3.functor(x);
        base.defined = defined(errorLow, errorHigh, xValue, yValue);
        return base;
    };
</code>

Can you see a better solution?

**Comment by ColinEberhardt** [comment]
Very good point ...

> Maybe we should make it another issue / PR?

Yes, totally agree, pull those changes out of this PR so that it can get merged. Thanks :-)

**Comment by bjedrzejewski** [comment]
It works if we use it like this:

`base.defined = function(d, i) {
        return defined(d, i, xValue, yOpenValue, yLowValue, yHighValue, yCloseValue);
    };`


I'm happy to commit that.

**Comment by ColinEberhardt** [comment]
> It works if we use it like this:

:+1: from me

**Comment by bjedrzejewski** [comment]
@ColinEberhardt ready to merge then!

**Comment by ColinEberhardt** [comment]
> ready to merge then!

Sorry - not quite ... see the comment above.

**Comment by bjedrzejewski** [comment]
@ColinEberhardt I updated to the style that we agreed. Sorry I have got confused in the discussion a bit.

**Comment by ColinEberhardt** [comment]
no worries - still one final tweak needed ;-)

**Comment by bjedrzejewski** [comment]
I removed that check :)

**Comment by ColinEberhardt** [comment]
I've squashed the commits and added a visual test ... see the branch here:

https://github.com/ColinEberhardt/d3-financial-components/commits/bjedrzejewski-boxplot

I do think I have found an issue. If you use my visual test, sometimes the chart renders boxes where the whiskers fall outside of the y axis range:

image

If you look at the screenshot above, the data for the highlighted box doesn't seem to match with the rendered view. Is there an issues with your SVG path logic?

**Comment by bjedrzejewski** [comment]
It definitely looks suspicious. I will investigate tomorrow.

**Comment by chrisprice** [comment]
I've landed this directly onto master (e743e0f70ddf6730dc8edc29ed730710ca2e33d8). I've additionally fixed the bug and simplified the implementation of the errorBar and boxPlot series (b22a125c27aea560e314e1f17a8f5430e080895f & f44c859d5d095e34d14037b0754182e43dfc46b6).

Thanks @bjedrzejewski!

**Pull request closed without merging**